File: src/test/java/org/reflections/ReflectionUtilsTest.java
package org.reflections;

import com.google.common.base.Function;
import com.google.common.collect.Sets;
import org.hamcrest.BaseMatcher;
import org.hamcrest.Description;
import org.junit.Test;
import org.reflections.TestModel.*;
import org.reflections.scanners.FieldAnnotationsScanner;

import java.lang.annotation.Annotation;
import java.lang.reflect.*;
import java.util.*;

import static com.google.common.collect.Collections2.transform;
import static org.junit.Assert.*;
import static org.reflections.ReflectionUtils.*;
import static org.reflections.ReflectionsTest.are;

/**
 * @author mamo
 */
@SuppressWarnings("unchecked")
public class ReflectionUtilsTest {

    @Test
    public void getAllTest() {
        assertThat(getAllSuperTypes(C3.class, withAnnotation(AI1.class)), are(I1.class));

        Set<Method> allMethods  = getAllMethods(C4.class, withModifier(Modifier.PUBLIC), withReturnType(void.class));
        Set<Method> allMethods1 = getAllMethods(C4.class, withPattern("public.*.void .*"));

        assertTrue(allMethods.containsAll(allMethods1) && allMethods1.containsAll(allMethods));
        assertThat(allMethods1, names("m1"));

        assertThat(getAllMethods(C4.class, withAnyParameterAnnotation(AM1.class)), names("m4"));

        assertThat(getAllFields(C4.class, withAnnotation(AF1.class)), names("f1", "f2"));

        assertThat(getAllFields(C4.class, withAnnotation(new AF1() {
            @Override
            public String value() {return "2";}

            @Override
            public Class<? extends Annotation> annotationType() {return AF1.class;}
        })), names("f2"));

        assertThat(getAllFields(C4.class, withTypeAssignableTo(String.class)), names("f1", "f2", "f3"));

        assertThat(getAllConstructors(C4.class, withParametersCount(0)), names(C4.class.getName()));

        assertEquals(getAllAnnotations(C3.class).size(), 5);

        Method m4 = getMethods(C4.class, withName("m4")).iterator().next();
        assertEquals(m4.getName(), "m4");
        assertTrue(getAnnotations(m4).isEmpty());
    }

    @Test
    public void withParameter() throws Exception {
        Class  target = Collections.class;
        Object arg1   = Arrays.asList(1, 2, 3);

        Set<Method> allMethods = Sets.newHashSet();
        for (Class<?> type : getAllSuperTypes(arg1.getClass())) {
            allMethods.addAll(getAllMethods(target, withModifier(Modifier.STATIC), withParameters(type)));
        }

        Set<Method> allMethods1 = getAllMethods(target,
                                                withModifier(Modifier.STATIC),
                                                withParametersAssignableTo(arg1.getClass()));

        assertEquals(allMethods, allMethods1);

        for (Method method : allMethods) { //effectively invokable
            Object invoke = method.invoke(null, arg1);
        }
    }

    @Test
    public void withParametersAssignableFromTest() throws Exception {
        //Check for null safe
        getAllMethods(Collections.class, withModifier(Modifier.STATIC), withParametersAssignableFrom());

        Class  target = Collections.class;
        Object arg1   = Arrays.asList(1, 2, 3);

        Set<Method> allMethods = Sets.newHashSet();
        for (Class<?> type : getAllSuperTypes(arg1.getClass())) {
            allMethods.addAll(getAllMethods(target, withModifier(Modifier.STATIC), withParameters(type)));
        }

        Set<Method> allMethods1 = getAllMethods(target,
                                                withModifier(Modifier.STATIC),
                                                withParametersAssignableFrom(Iterable.class),
                                                withParametersAssignableTo(arg1.getClass()));

        assertEquals(allMethods, allMethods1);

        for (Method method : allMethods) { //effectively invokable
            Object invoke = method.invoke(null, arg1);
        }
    }

    @Test
    public void withReturn() {
        Set<Method> returnMember              = getAllMethods(Class.class, withReturnTypeAssignableTo(Member.class));
        Set<Method> returnsAssignableToMember = getAllMethods(Class.class, withReturnType(Method.class));

        assertTrue(returnMember.containsAll(returnsAssignableToMember));
        assertFalse(returnsAssignableToMember.containsAll(returnMember));

        returnsAssignableToMember = getAllMethods(Class.class, withReturnType(Field.class));
        assertTrue(returnMember.containsAll(returnsAssignableToMember));
        assertFalse(returnsAssignableToMember.containsAll(returnMember));
    }

    @Test
    public void getAllAndReflections() {
        Reflections reflections = new Reflections(TestModel.class, new FieldAnnotationsScanner());

        Set<Field> af1       = reflections.getFieldsAnnotatedWith(AF1.class);
        Set<Field> allFields = getAll(af1, withModifier(Modifier.PROTECTED));
        assertTrue(allFields.size() == 1);
        assertThat(allFields, names("f2"));
    }

    private Set<String> names(Set<? extends Member> o) {
        return Sets.newHashSet(transform(o, (Function<Member, String>) input -> input.getName()));
    }

    private BaseMatcher<Set<? extends Member>> names(String... namesArray) {
        return new BaseMatcher<Set<? extends Member>>() {

            @Override
            public boolean matches(Object o) {
                Collection<String> transform = names((Set<Member>) o);
                Collection<?>      names     = Arrays.asList(namesArray);
                return transform.containsAll(names) && names.containsAll(transform);
            }

            @Override
            public void describeTo(Description description) {
            }
        };
    }
}

File: src/test/java/org/reflections/ReflectionUtilsTest.kt
package org.reflections

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.reflections.ReflectionUtils.getAllMethods
import org.reflections.ReflectionUtils.withModifier
import org.reflections.ReflectionUtils.withParameters
import org.reflections.ReflectionUtils.withParametersAssignableFrom
import org.reflections.ReflectionUtils.withParametersAssignableTo
import org.reflections.TestModel.AF1
import org.reflections.TestModel.AI1
import org.reflections.TestModel.AM1
import org.reflections.TestModel.C3
import org.reflections.TestModel.C4
import org.reflections.TestModel.I1
import org.reflections.scanners.FieldAnnotationsScanner
import org.reflections.util.classAndInterfaceHieararchyExceptObject
import org.reflections.util.urlForClass
import org.reflections.util.withAnnotation
import org.reflections.util.withAnyParameterAnnotation
import java.lang.reflect.Field
import java.lang.reflect.Member
import java.lang.reflect.Method
import java.lang.reflect.Modifier
import java.util.*
import java.util.regex.Pattern

/**
 * @author mamo
 */
class ReflectionUtilsTest {
    @Test
    fun getAllTest() {
        assertEquals(listOf(I1::class.java),
                     C3::class.java.classAndInterfaceHieararchyExceptObject().filter { it.isAnnotationPresent(AI1::class.java) })

        val allMethods =
                getAllMethods(C4::class.java).filter(andPredicate({ withModifier(it, Modifier.PUBLIC) },
                                                                  { it.returnType == Void.TYPE })).toSet()
        val allMethods1 = getAllMethods(C4::class.java).filter { it: Method ->
            Pattern.matches("public.*.void .*", it.toString())
        }.toSet()

        assertTrue(allMethods.containsAll(allMethods1) && allMethods1.containsAll(allMethods))
        assertHasNames(listOf("m1", "m1"), allMethods1)

        assertHasNames(listOf("m4"), getAllMethods(C4::class.java).filter { it: Method ->
            withAnyParameterAnnotation(it, AM1::class.java)
        }.toSet())

        assertHasNames(listOf("f1", "f2"),
                       ReflectionUtils.getAllFields(C4::class.java) { it.isAnnotationPresent(AF1::class.java) })

        assertHasNames(listOf("f2"), ReflectionUtils.getAllFields(C4::class.java) {
            withAnnotation(it, JavaSpecific.newAF1("2"))
        })

        assertHasNames(listOf("f1", "f2", "f3"), ReflectionUtils.getAllFields(C4::class.java) {
            String::class.java.isAssignableFrom(it.type)
        })

        assertEquals(ReflectionUtils.getAllAnnotations(C3::class.java).size.toLong(), 5)

        val m4 = C4::class.java.declaredMethods.filter { it.name == "m4" }.iterator().next()
        assertEquals("m4", m4.name)
        assertTrue(ReflectionUtils.getAnnotations(m4).isEmpty())
    }

    @Test
    fun withParameter() {
        val target = Collections::class.java
        val arg1 = listOf(1, 2, 3)

        val allMethods = mutableSetOf<Method>()
        for (type in arg1.javaClass.classAndInterfaceHieararchyExceptObject()) {
            allMethods.addAll(getAllMethods(target).filter(andPredicate({ withModifier(it, Modifier.STATIC) },
                                                                        { withParameters(it, listOf(type)) })).toSet())
        }

        val allMethods1 =
                getAllMethods(target).filter(andPredicate({ withModifier(it, Modifier.STATIC) },
                                                          { withParametersAssignableTo(it, listOf(arg1.javaClass)) }))
                    .toSet()

        assertEquals(allMethods1, allMethods)

        for (method in allMethods) { //effectively invokable
            val invoke = method.invoke(null, arg1)
        }
    }

    @Test
    fun withParametersAssignableFromTest() {
        //Check for null safe
        getAllMethods(Collections::class.java).filter(andPredicate({ withModifier(it, Modifier.STATIC) },
                                                                   { withParametersAssignableFrom(it, emptyList()) }))
            .toSet()

        val target = Collections::class.java
        val arg1 = listOf(1, 2, 3)

        val allMethods = mutableSetOf<Method>()
        for (type in arg1.javaClass.classAndInterfaceHieararchyExceptObject()) {
            allMethods.addAll(getAllMethods(target).filter(andPredicate({ withModifier(it, Modifier.STATIC) },
                                                                        { withParameters(it, listOf(type)) })).toSet())
        }

        val allMethods1 = getAllMethods(target).filter(andPredicate({ withModifier(it, Modifier.STATIC) }, {
            withParametersAssignableFrom(it, listOf(Iterable::class.java))
        }, { withParametersAssignableTo(it, listOf(arg1.javaClass)) })).toSet()

        assertEquals(allMethods, allMethods1)

        for (method in allMethods) { //effectively invokable
            val invoke = method.invoke(null, arg1)
        }
    }

    @Test
    fun withReturn() {
        val returnMember =
                getAllMethods(Class::class.java).filter { it: Method -> Member::class.java.isAssignableFrom(it.returnType) }
                    .toSet()
        var returnsAssignableToMember =
                getAllMethods(Class::class.java).filter { it: Method -> it.returnType == Method::class.java }.toSet()

        assertTrue(returnMember.containsAll(returnsAssignableToMember))
        assertFalse(returnsAssignableToMember.containsAll(returnMember))

        returnsAssignableToMember =
                getAllMethods(Class::class.java).filter { it: Method -> it.returnType == Field::class.java }.toSet()
        assertTrue(returnMember.containsAll(returnsAssignableToMember))
        assertFalse(returnsAssignableToMember.containsAll(returnMember))
    }

    @Test
    fun getAllAndReflections() {
        val reflections =
                Reflections(Configuration(filter = Filter.Include(TestModel::class.java.toPackageNameRegex()),
                                          urls = listOfNotNull(urlForClass(TestModel::class.java)).toSet(),
                                          scanners = setOf(FieldAnnotationsScanner())))

        val allFields =
                reflections.getFieldsAnnotatedWith(AF1::class.java).filter { withModifier(it, Modifier.PROTECTED) }
                    .toSet()
        assertEquals(1, allFields.size.toLong())
        assertHasNames(listOf("f2"), allFields)
    }

    private fun assertHasNames(expected: List<String>, members: Iterable<Member>) =
            assertEquals(expected, members.map { it.name })

    private fun <T> andPredicate(vararg predicates: (T) -> Boolean): (T) -> Boolean =
            { input -> predicates.all { predicate -> predicate(input) } }
}
